### PR TITLE
[TRTLLM-15447][feat] Add model startup stage timing metrics

### PR DESCRIPTION
**Pull request by yijingl-nvidia**
## Summary

Add wall-clock startup timing breakdowns for PyExecutor construction and initial/final model-engine warmup, exposed through `LLM.startup_metrics` and `GET /server_info`.

New timing metrics:

PyExecutor construction
- config_and_checkpoint_loader_initialization_seconds
- model_engine_creation_seconds
- draft_model_engine_creation_seconds
- guided_decoder_creation_seconds
- sampler_creation_seconds
- initial_kv_cache_creation_seconds
- speculative_decoding_resource_manager_creation_seconds
- speculative_drafter_creation_seconds
- initial_py_executor_creation_seconds_for_kv_cache_estimation
- kv_cache_capacity_configuration_seconds
- final_kv_cache_creation_seconds
- final_py_executor_creation_seconds
- worker_start_seconds
- total_py_executor_creation_seconds

Model-engine warmup (initial_model_engine / final_model_engine)
- sampling_warmup_seconds
- attention_warmup_seconds
- general_warmup_seconds
- autotuner_warmup_seconds
- mamba_hybrid_warmup_seconds
- gen_cuda_graph_warmup_seconds
- gen_cuda_graph_capture_seconds
- ctx_cuda_graph_warmup_seconds
- ctx_cuda_graph_capture_seconds
- post_ctx_cuda_graph_capture_warmup_seconds
- dg_paged_mqa_warmup_seconds
- cute_dsl_radix_topk_warmup_seconds
- memory_pool_prepopulation_seconds
- kv_cache_cleanup_seconds
- total_warmup_seconds

## Testing
- Added six CPU-only tests.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## Dev Engineer Review

- Adds startup timing metrics for `PyExecutor` creation and model-engine warmup.
- Exposes metrics through `LLM.startup_metrics` and `/server_info`.
- Separates sampling, CUDA graph warmup, and CUDA graph capture timings.
- Transfers initial and final model-engine metrics into startup metrics and clears source metrics.
- Records executor-creation duration when creation raises an error.
- Updates API documentation and replaces the piecewise warmup threshold literal with a named constant.
- Review finding counts and test execution results were not provided.

## QA Engineer Review

- Adds and updates unit tests for executor creation, model-engine warmup, and `BaseWorker` startup metrics.
- Covers timing scopes, CUDA graph stages, cleanup failures, disabled capture, metric transfer and filtering, and executor-creation errors.
- `tests/unittest/executor/test_base_worker.py` is listed in `tests/integration/test_lists/test-db/l0_cpu.yml` and `l0_a100.yml`.
- No matching test-list entries were found for the other changed unit tests.
- **Coverage verdict: needs follow-up.**

## Per-File QA Perspective

- `docs/source/llm-api/index.md`: Verify documented metric names, scopes, conditional stages, and example values.
- `tensorrt_llm/_torch/compilation/piecewise_optimizer.py`: Verify that the named threshold preserves behavior.
- `tensorrt_llm/_torch/pyexecutor/model_engine.py`: Verify warmup scopes, graph timing separation, cleanup, and `metrics`.
- `tensorrt_llm/_torch/pyexecutor/py_executor.py`: Verify the new `metrics` property and structure.
- `tensorrt_llm/_torch/pyexecutor/py_executor_creator.py`: Verify stage totals, metric transfer, and source clearing.
- `tensorrt_llm/executor/base_worker.py`: Verify startup metric classification and promotion.
- `tensorrt_llm/_torch/pyexecutor/breakable_cuda_graph_runner.py`: Verify separate warmup and capture timings and synchronization.
- `tests/unittest/_torch/executor/test_py_executor_creator_flash_mla_tokens_per_block.py`: Verifies `_create_py_executor`; no matching test-list entry was found.
- `tests/unittest/_torch/executor/test_py_executor_creator_mla_cache_reuse_sync.py`: Covers stage names, transfer, clearing, and error timing; no matching test-list entry was found.
- `tests/unittest/_torch/executor/test_pytorch_model_engine_warmup.py`: Covers warmup scopes, graph stages, cleanup failures, and disabled capture; no matching test-list entry was found.
- `tests/unittest/executor/test_base_worker.py`: Covers startup metric promotion and filtering. It is listed in `l0_cpu.yml` and `l0_a100.yml`.

### CODE REVIEW

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 1]
```diff
@@ -1475,11 +1493,14 @@ def warmup(self, resource_manager: ResourceManager) -> None:
             with self.cuda_graph_runner.allow_capture():
                 self.cuda_graph_runner.is_warmup_only = True
                 try:
-                    self._run_cuda_graph_warmup(resource_manager)
+                    with timing_metric("cuda_graph_warmup_seconds",
+                                       self._metrics):
+                        self._run_cuda_graph_warmup(resource_manager)
```
Comment (chienchunhung) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py:1498]: Please preserve current main’s `maybe_autotune_lora()` context when resolving this conflict. It owns both LoRA CUDA-graph autotuning and the PP cache receive/send/cleanup in its finally block. Please nest it inside `timing_metric("cuda_graph_warmup_seconds", self._metrics)` so the metric covers the full stage without dropping the behavior added by #17412.

File: docs/source/llm-api/index.md [hunk 2]
```diff
@@ -63,27 +63,113 @@ A typical result has the following structure:
...
+  },
   "model_loader": {
-    "total_model_loading_seconds":  1.971,
-    "checkpoint_preparation_seconds": 1.177,
-    "weight_population_seconds": 0.598,
-    "post_load_processing_seconds": 0.005
+    "checkpoint_preparation_seconds": 0.719,
+    "weight_population_seconds": 0.272,
+    "post_load_processing_seconds": 0.004,
+    "total_model_loading_seconds": 1.545
```
Comment (chienchunhung) [inline docs/source/llm-api/index.md:110]: Current main reports `checkpoint_finalization_seconds` and `draft_checkpoint_finalization_seconds`, added by #17393. Please retain those fields in the example, prose, and metric table when resolving this conflict so the public startup-metrics documentation continues to match the payload.

**Review by chienchunhung — COMMENTED** [review]
Thanks for the PR; nice work!

Some high-level suggesitons in addition to the inline comments:

`sampling_warmup` currently appears only within `total_warmup_seconds`, while `piecewise_cuda_graph_warmup` and `piecewise_cuda_graph_capture` are folded into `cuda_graph_capture_seconds`. This prevents users from isolating several significant cold-start contributors—especially sampling warmup, which the investigation measured at roughly 70 seconds.

Could we add dedicated metrics for `sampling_warmup_seconds`, `piecewise_cuda_graph_warmup_seconds`, and `piecewise_cuda_graph_capture_seconds` under each applicable model-engine stage? That would make the exported metrics actionable and allow the investigation’s startup breakdown to be reproduced.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 3]
```diff
@@ -1475,11 +1493,14 @@ def warmup(self, resource_manager: ResourceManager) -> None:
...
                 finally:
                     self.cuda_graph_runner.is_warmup_only = False
                 self.cuda_graph_runner.padding_dummy_requests = {}
-                self._run_cuda_graph_warmup(resource_manager)
+                with timing_metric("cuda_graph_capture_seconds", self._metrics):
```
Comment (AlessioNetti) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py:1502]: Note that the range above encloses both generation and piecewise CUDA graph capture - in case you'd like to split this in separate `piecewise_cuda_graph_capture/warmup_seconds` and `cuda_graph_capture_seconds` timings, you'll need to move the `timing_metric` block inside of `_run_cuda_graph_warmup`, and account for the value of `is_warmup_only`. Piecewise CUDA graph capture and warmup can be measured separately inside `_capture_prefill_cuda_graphs`.

**Comment by AlessioNetti** [comment]
Thank you for the changes - I had a chance to take a look at the code and I agree with @chienchunhung's assessment, especially that it would be useful to have a `sampling_warmup_seconds` timing, as well as distinct timings for piecewise CUDA graph capture & warmup. I left a small comment with some suggestions for the latter :)

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 1 above)
Comment (yijingl-nvidia) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: Addressed in the latest PR version

**Comment by yijingl-nvidia** [comment]
> Thanks for the PR; nice work!
> 
> Some high-level suggesitons in addition to the inline comments:
> 
> `sampling_warmup` currently appears only within `total_warmup_seconds`, while `piecewise_cuda_graph_warmup` and `piecewise_cuda_graph_capture` are folded into `cuda_graph_capture_seconds`. This prevents users from isolating several significant cold-start contributors—especially sampling warmup, which the investigation measured at roughly 70 seconds.
> 
> Could we add dedicated metrics for `sampling_warmup_seconds`, `piecewise_cuda_graph_warmup_seconds`, and `piecewise_cuda_graph_capture_seconds` under each applicable model-engine stage? That would make the exported metrics actionable and allow the investigation’s startup breakdown to be reproduced.

New version of the PR broke down the warmup to isolate sampling warm up and prefill cuda graph warmup.

File: docs/source/llm-api/index.md (hunk 2 above)
Comment (yijingl-nvidia) [reply docs/source/llm-api/index.md:113]: Addressed

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 3 above)
Comment (yijingl-nvidia) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: Addressed on the latest PR version

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 4]
```diff
@@ -1403,7 +1417,8 @@ def warmup(self,
         # CUDA graph capture pass exercises the non-greedy sampler, so with
         # cuda_graph_config=None flashinfer's sampling kernels would be
         # JIT-built mid-serving.
-        warmup_sampling_module()
+        with timing_metric("sampling_warmup_seconds", self._metrics):
+            warmup_sampling_module()
```
Comment (chienchunhung) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py:1458]: `sampling_warmup_seconds` currently ends before `warmup_sample_from_logits_op()`. With in-graph sampling enabled, that helper would `torch.compile` the sampling operation for every configured graph batch size, so a potentially meaningful sampling startup cost is omitted from `sampling_warmup_seconds`. 

By the way, when resolving the current-main conflict, please preserve `_eager_workspace_reclaimer = None` and include both sampling warmup helpers in this timing scope. We can also add a focused test to cover the `enable_in_graph_sampling=True` path.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 5]
```diff
@@ -2927,62 +2966,68 @@ def _capture_prefill_cuda_graphs(self, resource_manager: ResourceManager):
         capture_context = (capture_piecewise_cuda_graph(True)
                            if self._torch_compile_piecewise_cuda_graph else
                            contextlib.nullcontext())
-        with capture_context, self.no_cuda_graph():
+        with timing_metric("ctx_cuda_graph_capture_seconds", self._metrics):
```
Comment (chienchunhung) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: IIRC `ctx_cuda_graph_capture_seconds` here still combines **pre-capture warmup** and **capture** together. 

For the piecewise backend, the first 3 forwards increment `warmup_count` and only the 4th captures; `BreakableCUDAGraphRunner.capture()` likewise runs `warmup()` before capture. `post_ctx_cuda_graph_capture_warmup_seconds` measures a different, later logits-buffer warmup, so the investigation’s warmup and capture phases remain unrecoverable. 

Could we record pre-capture context-graph warmup and capture separately and add a test asserting which forwards fall into each timing scope?

For reference, here's a possible idea suggested by Codex:

```python3
with timing_metric("ctx_cuda_graph_warmup_seconds", self._metrics):
    for _ in range(PIECEWISE_CUDA_GRAPH_WARMUP_STEPS):  # currently 3
        forward()
    torch.cuda.synchronize()

with timing_metric("ctx_cuda_graph_capture_seconds", self._metrics):
    forward()  # capture invocation, including initial replay
    torch.cuda.synchronize()
```

**Review by chienchunhung — APPROVED** [review]
Thanks for working on my earlier comments! I added 2 follow-up comments. The PR looks good to me once the comments are addressed and the conflicts are resolved.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 5 above)
Comment (yijingl-nvidia) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: Followed the comment and updated the PR

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 4 above)
Comment (yijingl-nvidia) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:1458]: Updated the PR according to the comment.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 6]
```diff
@@ -1279,17 +1287,19 @@ def wrapper(self,
...
-            if kv_cache_manager is not None:
-                has_invalid_values = kv_cache_manager.check_invalid_values_in_kv_cache(
-                    fill_with_zero=True)
-                if has_invalid_values:
-                    logger.warning(
-                        "NaNs/Infs have been introduced to KVCache during warmup, KVCache was filled with zeros to avoid potential issues"
-                    )
+            with timing_metric("total_warmup_seconds", self._metrics):
+                result = method(self, resource_manager, *args, **kwargs)
+                with timing_metric("kv_cache_cleanup_seconds", self._metrics):
```
[the remark this answers is not in the document]
Comment (yijingl-nvidia) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:1292]: Resolved in newest version of the PR

File: tensorrt_llm/_torch/pyexecutor/py_executor_creator.py [hunk 7]
```diff
@@ -373,10 +414,36 @@ def _create_py_executor_impl(
...
+    with timing_metric("total_py_executor_creation_seconds", creation_metrics):
+        py_executor = _create_py_executor(
+            llm_args=llm_args,
+            creation_metrics=creation_metrics,
+            checkpoint_dir=checkpoint_dir,
+            tokenizer=tokenizer,
+            profiling_stage_data=profiling_stage_data,
+            resource_governor_queue=resource_governor_queue,
+        )
+    py_executor.metrics.update(creation_metrics)
```
[the remark this answers is not in the document]
Comment (yijingl-nvidia) [reply tensorrt_llm/_torch/pyexecutor/py_executor_creator.py:427]: Resolved in newest version of the PR

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 8]
```diff
@@ -1057,6 +1058,11 @@ def on_detected():
         if start_worker:
             self.start_worker()
 
+    @property
+    def metrics(self) -> dict[str, float | dict[str, float]]:
+        """Return executor construction and model-engine startup metrics."""
+        return self._metrics
```
[the remark this answers is not in the document]
Comment (yijingl-nvidia) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py:1064]: Resolved in newest version of the PR

**Review by brnguyen2 — APPROVED** [review]
